### PR TITLE
feat(core): Add an NgZone subclass that makes OnPush Angular apps wor…

### PR DESCRIPTION
**Pull request by austinrmiller**
…k without Zone.js

Zone.js adds performance overhead and is not compatible (without source transformations) with native
async/await. For apps that consistently use OnPush, Zone.js can be treated as optional if the signal
that a view is dirty is connected to a mechanism to schedule change detection, which is what this
change does. This change also removes a conflicting test that was testing behavior that should not
affect correctness or performance.

Opt-in solution for #38782

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #38782 

Right now, NgZone is all-or-nothing with regard to automated change detection.

## What is the new behavior?

This adds a new NgZone implementation that does not depend on Zone.js and connects the code that marks views dirrty to the scheduling of change detection. For apps that consistently use OnPush change detection, this new zone should "just work" with no further changes in the application code. Even for apps that don't, "most" components will likely "just work" in this mode.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/interfaces/view.ts [hunk 1]
```diff
@@ -326,6 +327,11 @@ export interface LView extends Array<any> {
    * are not `Dirty`/`CheckAlways`.
    */
   [TRANSPLANTED_VIEWS_TO_REFRESH]: number;
+
+  /**
+   * Function to mark the view as dirty and ensure change dection is scheduled.
```
Comment (dario-piotrowicz) [inline packages/core/src/render3/interfaces/view.ts:332]:
```suggestion
   * Function to mark the view as dirty and ensure change detection is scheduled.
```

File: packages/core/src/view/util.ts [hunk 2]
```diff
@@ -117,13 +118,28 @@ export function markParentViewsForCheck(view: ViewData) {
...
   let currView: ViewData|null = view;
   while (currView && currView !== endView) {
     currView.state |= ViewState.CheckProjectedViews;
     currView = currView.viewContainerParent || currView.parent;
   }
 }
 
+export function detectChanges(view: ViewData): void {
+  const fs = view.root.rendererFactory;
+  if (fs.begin) {
```
Comment (dario-piotrowicz) [inline packages/core/src/view/util.ts:131]: here and below, couldn't you use optional chaining again as in `fs.being?.()` ?

Comment (austinrmiller) [reply packages/core/src/view/util.ts]: Done

File: packages/core/src/render3/interfaces/view.ts (hunk 1 above)
Comment (austinrmiller) [reply packages/core/src/render3/interfaces/view.ts]: Done

**Review by petebacondarwin — COMMENTED** [review]
I am a bit concerned about the code size increase for scenarios that do not use the `OnPushNgZone`. What do people think about making the `OnPushNgZone` class a public API instead?

File: packages/core/src/application_ref.ts [hunk 3]
```diff
@@ -457,12 +458,14 @@ export class PlatformRef {
...
 function getNgZone(
-    ngZoneOption: NgZone|'zone.js'|'noop'|undefined,
+    ngZoneOption: NgZone|'zone.js'|'noop'|'onpush'|undefined,
     extra?: {ngZoneEventCoalescing: boolean, ngZoneRunCoalescing: boolean}): NgZone {
   let ngZone: NgZone;
 
   if (ngZoneOption === 'noop') {
     ngZone = new NoopNgZone();
+  } else if (ngZoneOption === 'onpush') {
+    ngZone = new OnPushNgZone();
```
[the remark this answers is not in the document]
Comment (austinrmiller) [reply packages/core/src/application_ref.ts]: Sure, just trying to follow the pattern. Done.

File: packages/core/src/render3/component_ref.ts [hunk 4]
```diff
@@ -161,6 +163,12 @@ export class ComponentFactory<T> extends viewEngine_ComponentFactory<T> {
     const rootLView = createLView(
         null, rootTView, rootContext, rootFlags, null, null, rendererFactory, hostRenderer,
         sanitizer, rootViewInjector);
+    const ngZone = rootViewInjector.get(NgZone, {});
+    if (ngZone.markForCheck) {
+      rootLView[MARK_DIRTY] = () => {
+        ngZone.markForCheck?.();
```
[the remark this answers is not in the document]
Comment (austinrmiller) [reply packages/core/src/render3/component_ref.ts]: Done

**Comment by austinrmiller** [comment]
I'm not sure why this is increasing core/todo/bundle by 14k - any pointers there? I'm assuming it has something to do with the extra symbols being found in the bundle, but I'm also not sure why that would be happening.

**Comment by austinrmiller** [comment]
Hm, I suppose because compont_ref.ts now has a direct dep on NgZone whereas the todo app doesn't do the "normal" angular initialization but instead calls ɵrenderComponent directly. So then should the size increase be allowed or should I add something like an InjectionToken to add some indirection to prevent a direct dep?

**Comment by jessicajaniuk** [comment]
@austinrmiller, this has a bunch of file conflicts. Could you please rebase? I'll take a look after that.

**Comment by austinrmiller** [comment]
Done

**Comment by nvaralakshmi** [comment]
@austinrmiller It looks like this PR is outdated, please rebase the PR.

**Comment by alxhub** [comment]
Closing as we don't want to go in this direction (as discussed offline with @austinrmiller)

**Pull request closed without merging**